source/adapters/level_zero/v2/event.hpp

```diff
@@ -98,13 +98,14 @@ struct ur_event_handle_t_ : _ur_object {
   uint64_t getEventEndTimestamp();
 
 protected:
+  // Do not use ref couting on context to avoid circular dependency.
```

Comment:
do we need something like a weak pointer? You added quite a few comments like that. It would be better to enforce this using a type.

Reply:
I guess I could just do an alias for this, and add some explanation in the comment above that alias definition.

Reply:
My suggestion would be to create a weak<> type that has deleted copy constructors/operators and can only be created from a ref-counted ptr (using something like .downgrade() or whatever).

Reply:
Yeah, but in some situations there might be no rc<> instance. For example, when you create an eventPool in the context constructor, you need to pass ur_context_handle_t (this) to the eventPool. If it would only accept weak<> and if weak<> can only be created from rc<>, you would need a temporary rc<>. This is a bit verbose but on the other hand, you can use rc_val_only (to do ref count validation) so this could provider us with some more checks.

See the latest commit for how it would look like. Also, I guess if we decide to do that, then probably all handles in v2 should be replaced with rc<> rc_val_only<> or weak<>.

Reply:
Having weak<> could also allow us to check the pointers for NULL (only allow rc<> and weak<> to be != NULL). If we ever need the value to be null (2 places right now), we can use optional (see latest commit).

Reply:
I think I'd prefer to have a weak be nullable, like normal std::weak. So that when the original rc pointer goes out of scope, we can detect that.

> For example, when you create an eventPool in the context constructor, you need to pass ur_context_handle_t (this) to the eventPool.

That's a scenario for the `rc_val_only` type, because context must outlive a queue/command buffer associated with the eventPool. I think weak is for when we create objects that might possibly outlive their "parent" objects.

Reply:
> That's a scenario for the rc_val_only type, because context must outlive a queue/command buffer associated with the eventPool. I think weak is for when we create objects that might possibly outlive their "parent" objects.

Yeah, but the problem is that if you use `rc_val_only` then it will always fail during `release` because of how urContextRelease (and other functions are implemented) - we first decrement the RefCount (to 0) and only then call a destructor. We could implement a real shared_ptr like in c++ (where the control block is in separate allocation) so we could change the order of those ops but I don't know if it's worth it :d

Reply:
Hm, but on the other hand, probably every use case of rc_val_only will be like this so we could just check if the ref count is >= 0

Reply:
Tbh, I think I'll remove `rc_val_only`, as it's only applicable to device and platforms and there's a small chance we will have lifetime issues with those. We could just add some checks to the device/platform classes if we really want to.